src/validators/date.rs

```diff
@@ -47,16 +47,12 @@ impl Validator for DateValidator {
         _definitions: &'data Definitions<CombinedValidator>,
         _recursion_guard: &'s mut RecursionGuard,
     ) -> ValResult<'data, PyObject> {
-        let date = match input.validate_date(extra.strict.unwrap_or(self.strict)) {
+        let strict = extra.strict.unwrap_or(self.strict);
+        let date = match input.validate_date(strict) {
             Ok(date) => date,
-            // if the date error was an internal error, return that immediately
-            Err(ValError::InternalErr(internal_err)) => return Err(ValError::InternalErr(internal_err)),
-            Err(date_err) => match self.strict {
-                // if we're in strict mode, we doing try coercing from a date
-                true => return Err(date_err),
-                // otherwise, try creating a date from a datetime input
-                false => date_from_datetime(input, date_err),
-            }?,
+            // if the error was a parsing error, in lax mode we allow datetimes at midnight
+            Err(line_errors @ ValError::LineErrors(..)) if !strict => date_from_datetime(input)?.ok_or(line_errors)?,
```

Comment:
With apologies I couldn't help golfing this line, with the result of modifying `date_from_datetime` a little 🙈